File: app/src/main/java/com/kickstarter/viewmodels/ProjectUpdatesViewModel.java
package com.kickstarter.viewmodels;

import android.util.Pair;

import com.kickstarter.libs.ActivityViewModel;
import com.kickstarter.libs.ApiPaginator;
import com.kickstarter.libs.Environment;
import com.kickstarter.libs.KoalaContext;
import com.kickstarter.libs.utils.BooleanUtils;
import com.kickstarter.libs.utils.ListUtils;
import com.kickstarter.models.Project;
import com.kickstarter.models.Update;
import com.kickstarter.services.ApiClientType;
import com.kickstarter.services.apiresponses.UpdatesEnvelope;
import com.kickstarter.ui.IntentKey;
import com.kickstarter.ui.activities.ProjectUpdatesActivity;

import java.util.List;

import androidx.annotation.NonNull;
import rx.Observable;
import rx.subjects.BehaviorSubject;
import rx.subjects.PublishSubject;

import static com.kickstarter.libs.rx.transformers.Transformers.combineLatestPair;
import static com.kickstarter.libs.rx.transformers.Transformers.takePairWhen;
import static com.kickstarter.libs.rx.transformers.Transformers.takeWhen;

public interface ProjectUpdatesViewModel {

  interface Inputs {
    /** Call when pagination should happen.*/
    void nextPage();

    /** Call when the feed should be refreshed. */
    void refresh();

    /** Call when an Update is clicked. */
    void updateClicked(Update update);
  }

  interface Outputs {
    /** Emits a boolean indicating whether the horizontal ProgressBar is visible. */
    Observable<Boolean> horizontalProgressBarIsGone();

    /** Emits a boolean indicating whether updates are being fetched from the API. */
    Observable<Boolean> isFetchingUpdates();

    /** Emits the current project and its updates. */
    Observable<Pair<Project, List<Update>>> projectAndUpdates();

    /** Emits a project and an update to start the update activity with. */
    Observable<Pair<Project, Update>> startUpdateActivity();
  }

  final class ViewModel extends ActivityViewModel<ProjectUpdatesActivity> implements Inputs, Outputs {
    private final ApiClientType client;

    public ViewModel(final @NonNull Environment environment) {
      super(environment);

      this.client = environment.apiClient();

      final Observable<Project> project = intent()
        .map(i -> i.getParcelableExtra(IntentKey.PROJECT))
        .ofType(Project.class)
        .take(1);

      final Observable<Project> startOverWith = Observable.merge(
        project,
        project.compose(takeWhen(this.refresh))
      );

      final ApiPaginator<Update, UpdatesEnvelope, Project> paginator =
        ApiPaginator.<Update, UpdatesEnvelope, Project>builder()
          .nextPage(this.nextPage)
          .startOverWith(startOverWith)
          .envelopeToListOfData(UpdatesEnvelope::updates)
          .envelopeToMoreUrl(env -> env.urls().api().moreUpdates())
          .loadWithParams(this.client::fetchUpdates)
          .loadWithPaginationPath(this.client::fetchUpdates)
          .clearWhenStartingOver(false)
          .concater(ListUtils::concatDistinct)
          .build();

      project
        .compose(combineLatestPair(paginator.paginatedData().share()))
        .compose(bindToLifecycle())
        .subscribe(this.projectAndUpdates);

      paginator
        .isFetching()
        .distinctUntilChanged()
        .take(2)
        .map(BooleanUtils::negate)
        .compose(bindToLifecycle())
        .subscribe(this.horizontalProgressBarIsGone);

      paginator
        .isFetching()
        .compose(bindToLifecycle())
        .subscribe(this.isFetchingUpdates);

      project
        .compose(takePairWhen(this.updateClicked))
        .compose(bindToLifecycle())
        .subscribe(this.startUpdateActivity::onNext);

      project
        .compose(takeWhen(this.updateClicked))
        .compose(bindToLifecycle())
        .subscribe(p -> this.koala.trackViewedUpdate(p, KoalaContext.Update.UPDATES));

      project
        .take(1)
        .compose(bindToLifecycle())
        .subscribe(this.koala::trackViewedUpdates);
    }

    private final PublishSubject<Void> nextPage = PublishSubject.create();
    private final PublishSubject<Void> refresh = PublishSubject.create();
    private final PublishSubject<Update> updateClicked = PublishSubject.create();

    private final BehaviorSubject<Boolean> horizontalProgressBarIsGone = BehaviorSubject.create();
    private final BehaviorSubject<Boolean> isFetchingUpdates = BehaviorSubject.create();
    private final BehaviorSubject<Pair<Project, List<Update>>> projectAndUpdates = BehaviorSubject.create();
    private final PublishSubject<Pair<Project, Update>> startUpdateActivity = PublishSubject.create();

    public final Inputs inputs = this;
    public final Outputs outputs = this;

    @Override public void nextPage() {
      this.nextPage.onNext(null);
    }
    @Override public void refresh() {
      this.refresh.onNext(null);
    }
    @Override public void updateClicked(final @NonNull Update update) {
      this.updateClicked.onNext(update);
    }

    @Override public @NonNull Observable<Boolean> horizontalProgressBarIsGone() {
      return this.horizontalProgressBarIsGone;
    }
    @Override public @NonNull Observable<Boolean> isFetchingUpdates() {
      return this.isFetchingUpdates;
    }
    @Override public @NonNull Observable<Pair<Project, List<Update>>> projectAndUpdates() {
      return this.projectAndUpdates;
    }
    @Override public @NonNull Observable<Pair<Project, Update>> startUpdateActivity() {
      return this.startUpdateActivity;
    }
  }
}

File: app/src/main/java/com/kickstarter/viewmodels/ProjectUpdatesViewModel.kt
package com.kickstarter.viewmodels

import android.content.Intent
import android.content.SharedPreferences
import android.util.Pair
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.kickstarter.libs.Environment
import com.kickstarter.libs.loadmore.ApolloPaginateV2
import com.kickstarter.libs.rx.transformers.Transformers
import com.kickstarter.libs.utils.EventContextValues
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.libs.utils.extensions.isNotNull
import com.kickstarter.libs.utils.extensions.storeCurrentCookieRefTag
import com.kickstarter.models.Project
import com.kickstarter.models.Update
import com.kickstarter.services.ApolloClientTypeV2
import com.kickstarter.services.apiresponses.updatesresponse.UpdatesGraphQlEnvelope
import com.kickstarter.ui.IntentKey
import com.kickstarter.ui.data.ProjectData
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.BehaviorSubject
import java.net.CookieManager

interface ProjectUpdatesViewModel {
    interface Inputs {
        /** Call when pagination should happen. */
        fun nextPage()

        /** Call when the feed should be refreshed.  */
        fun refresh()

        /** Call when an Update is clicked.  */
        fun updateClicked(update: Update)
    }

    interface Outputs {
        /** Emits a boolean indicating whether the horizontal ProgressBar is visible.  */
        fun horizontalProgressBarIsGone(): Observable<Boolean>

        /** Emits a boolean indicating whether updates are being fetched from the API.  */
        fun isFetchingUpdates(): Observable<Boolean>

        /** Emits the current project and its updates.  */
        fun projectAndUpdates(): Observable<Pair<Project, List<Update>>>

        /** Emits a project and an update to start the update activity with.  */
        fun startUpdateActivity(): Observable<Pair<Project, Update>>
    }

    class ProjectUpdatesViewModel(environment: Environment, private val intent: Intent? = null) :
        ViewModel(), Inputs, Outputs {
        private val client: ApolloClientTypeV2 = requireNotNull(environment.apolloClientV2())
        private val cookieManager: CookieManager = requireNotNull(environment.cookieManager())
        private val sharedPreferences: SharedPreferences = requireNotNull(environment.sharedPreferences())
        private val analyticEvents = requireNotNull(environment.analytics())
        private val nextPage = BehaviorSubject.create<Unit>()
        private val refresh = BehaviorSubject.create<Unit>()
        private val updateClicked = BehaviorSubject.create<Update>()
        private val horizontalProgressBarIsGone = BehaviorSubject.create<Boolean>()
        private val isFetchingUpdates = BehaviorSubject.create<Boolean>()
        private val projectAndUpdates = BehaviorSubject.create<Pair<Project, List<Update>>>()
        private val startUpdateActivity = BehaviorSubject.create<Pair<Project, Update>>()

        private val disposables = CompositeDisposable()

        val inputs: Inputs = this
        val outputs: Outputs = this

        private fun intent() = intent?.let { Observable.just(it) } ?: Observable.empty()
        init {

            val projectData = intent()
                .filter { (it.getParcelableExtra(IntentKey.PROJECT_DATA) as? ProjectData?).isNotNull() }
                .map { it.getParcelableExtra(IntentKey.PROJECT_DATA) as? ProjectData? }
                .filter { it.isNotNull() }
                .take(1)

            val project = projectData
                .filter { it.project().isNotNull() }
                .map { it.project() }

            projectData
                .map {
                    it.storeCurrentCookieRefTag(
                        cookieManager, sharedPreferences
                    )
                }
                .subscribe {
                    analyticEvents.trackProjectScreenViewed(
                        it, EventContextValues.ContextSectionName.UPDATES.contextName
                    )
                }
                .addToDisposable(disposables)

            val startOverWith = Observable.merge(
                project,
                project.compose(Transformers.takeWhenV2(refresh))
            )

            val pagination =
                ApolloPaginateV2.builder<Update, UpdatesGraphQlEnvelope, Project>()
                    .nextPage(nextPage)
                    .distinctUntilChanged(true)
                    .startOverWith(startOverWith)
                    .isReversed(false)
                    .envelopeToListOfData {
                        it.updates?.filter { updates ->
                            !updates.body().isNullOrEmpty()
                        }
                    }
                    .loadWithParams {
                        loadWithProjectUpdatesList(Observable.just(it.first), it.second)
                    }
                    .clearWhenStartingOver(false)
                    .build()

            pagination.isFetching
                .share()
                .distinctUntilChanged()
                .subscribe {
                    horizontalProgressBarIsGone.onNext(!it)
                    isFetchingUpdates.onNext(it)
                }
                .addToDisposable(disposables)

            val updatesList = pagination.paginatedData()?.share()?.filter { it.isNotNull() }?.map { requireNotNull(it) } ?: Observable.empty()

            project
                .compose<Pair<Project, List<Update>>>(Transformers.combineLatestPair(updatesList))
                .subscribe { projectAndUpdates.onNext(it) }
                .addToDisposable(disposables)

            project
                .compose(Transformers.takePairWhenV2(updateClicked))
                .subscribe { startUpdateActivity.onNext(it) }
                .addToDisposable(disposables)
        }

        override fun onCleared() {
            client.cleanDisposables()
            disposables.clear()
            super.onCleared()
        }
        private fun loadWithProjectUpdatesList(
            project: Observable<Project>,
            cursor: String
        ): Observable<UpdatesGraphQlEnvelope> {
            return project.switchMap {
                return@switchMap client.getProjectUpdates(it.slug() ?: "", cursor)
            }.onErrorResumeNext(Observable.empty())
        }
        override fun nextPage() {
            nextPage.onNext(Unit)
        }

        override fun refresh() {
            refresh.onNext(Unit)
        }

        override fun updateClicked(update: Update) {
            updateClicked.onNext(update)
        }

        override fun horizontalProgressBarIsGone(): Observable<Boolean> = horizontalProgressBarIsGone
        override fun isFetchingUpdates(): Observable<Boolean> = isFetchingUpdates
        override fun projectAndUpdates(): Observable<Pair<Project, List<Update>>> = projectAndUpdates
        override fun startUpdateActivity(): Observable<Pair<Project, Update>> = startUpdateActivity
    }

    class Factory(private val environment: Environment, private val intent: Intent) : ViewModelProvider.Factory {
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return ProjectUpdatesViewModel(environment, intent) as T
        }
    }
}
